File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.go
func letterCombinations(digits string) []string {
    
    if len(digits) == 0 {
        return []string{}
    }

    key := []string{
        "",    // 0
        "",    // 1
        "abc", // 2
        "def", // 3
        "ghi", // 4
        "jkl", // 5
        "mno", // 6
        "pqrs", // 7
        "tuv", // 8
        "wxyz", // 9
    }

    var ans []string
    var backtrack func(index int, path string)
    
    backtrack = func(index int, path string) {
        if index == len(digits) {
            ans = append(ans, path)
            return
        }

        digit := digits[index] - '0'
        possibleLetters := key[digit]

        for _, letter := range possibleLetters {
            backtrack(index+1, path+string(letter))
        }
    }

    backtrack(0, "")
    return ans
}

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str):
        res = []
        digitToChar = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "qprs",
            "8": "tuv",
            "9": "wxyz",
        }

        def backtrack(i, curStr):
            if len(curStr) == len(digits):
                res.append(''.join(curStr[:]))
                return
            for c in digitToChar[digits[i]]:
                curStr.append(c)
                backtrack(i + 1, curStr)
                curStr.pop()

        if digits:
            temp=[]
            backtrack(0, temp)

        return res
